SystemManager/Src/SM_States.cpp

```diff
@@ -0,0 +1,243 @@
+    // TODO Check if we received new PID tunings and call the AM function here
+
+    // Check all arming mechnisms triggered then move to GroundOp
+    // TODO
+
+    // Check switch here that changes modes
+    system_manager->rc_data = Los_Link::getInstance().getRx();
+    if (system_manager->rc_data.rx_channels[SM::RC_ARM_CHANNEL] > 90) {
+        system_manager->setState(GroundOpMode::getInstance());
+    }
+}
+
+SystemState& DisarmMode::getInstance() {
+    static DisarmMode singleton;
+    return singleton;
+}
+
+/********************
+ * Ground Op Mode
+ ********************/
+void GroundOpMode::execute(SystemManager *system_manager) {
+    // To move to flight check that armed switch is high,
+    // throttle and pitch are 0 (both sticks down)
+    system_manager->rc_data = Los_Link::getInstance().getRx();
+    if (system_manager->rc_data.rx_channels[SM::RC_ARM_CHANNEL] > 90 &&
+        system_manager->rc_data.rx_channels[SM::RC_THROTTLE_CHANNEL] == 0 &&
+        system_manager->rc_data.rx_channels[SM::RC_PITCH_CHANNEL]) {
+        system_manager->setState(FlightMode::getInstance());
+    }
+}
+
+void GroundOpMode::exit(SystemManager *system_manager) {
+    // Delete SM to AM queue to prevent PID tuning midair (not yet in Nov 27th ver)
+    // osMessageQueueReset(system_manager->SM_to_AM_queue);
+    // system_manager->SM_to_AM_queue = NULL;
+}
+
+SystemState& GroundOpMode::getInstance() {
+    static GroundOpMode singleton;
+    return singleton;
+}
+
+/********************
+ * Flight Mode
+ ********************/
+void FlightMode::enter(SystemManager *system_manager) {
+    // Start the PM thread and increase TM speed (have to start different task)
+    // xTaskCreate(system_manager->PMOperationTask, "PM Thread", &system_manager->PM_handle);
+    // vTaskDelete(system_manager->TM_handle);
+    // xTaskCreate(system_manager->TMOperationTask, "TM Thread", &system_manager->TM_handle);
+
+    // system_manager->SM_to_PM_queue = osMessageQueueNew(1, 256, NULL);
+    // system_manager->PM_to_AM_queue = osMessageQueueNew(1, 256, NULL);
+
+    // JUST for the Nov 27th deadline flight:
+
+    system_manager->SM_to_AM_queue = osMessageQueueNew(1U, 256U, NULL);
+    system_manager->attitude_manager.setSmQueue(system_manager->SM_to_AM_queue);
+    AM::AttitudeManager* am_pointer = &system_manager->attitude_manager;
+    xTaskCreate(system_manager->AMOperationTask, "AM Thread", 400U,
+                (void *)am_pointer, osPriorityNormal,
+                &system_manager->AM_handle);
+    // Task will automatically be added to the FreeRTOS scheduler
+}
+
+void FlightMode::execute(SystemManager *system_manager) {
+    // Get sensor fusion data from LOS interfaces
+    // TODO
+    // SM.sf_data = los_interface.
+
+    // Get RC data from LOS
+    system_manager->rc_data = Los_Link::getInstance().getRx();
+    if (system_manager->rc_data.rx_channels[SM::RC_ARM_CHANNEL] < 10) {
+        system_manager->setState(DisarmMode::getInstance());
```

Comment:
Are there any scenarios where we want to go to GroundOp mode from FlightMode? idk what that would look like, but I imagine after a landing we may want this?

Reply:
I think after landing that would make sense. Once landing is added to this it should be easy to add.